File: src/main/java/ru/spb/se/contexthelper/ui/StackExchangeThreadsTreeCellRender.java
package ru.spb.se.contexthelper.ui;

import com.google.code.stackexchange.schema.Answer;
import com.google.code.stackexchange.schema.Comment;
import com.google.code.stackexchange.schema.Question;
import com.google.common.annotations.VisibleForTesting;
import com.intellij.openapi.util.IconLoader;

import javax.swing.*;
import javax.swing.tree.TreeCellRenderer;
import java.awt.*;
import java.text.DecimalFormat;
import java.text.NumberFormat;

public class StackExchangeThreadsTreeCellRender implements TreeCellRenderer {

  private static final String QUESTION_ICON_PATH = "/icons/so-icon.png";
  private static final String ANSWER_ICON_PATH = "/icons/so-icon.png";

  @Override
  public Component getTreeCellRendererComponent(
      JTree tree, Object renderedObject, boolean selected, boolean expanded, boolean leaf,
      int row, boolean hasFocus) {
    JLabel label = new JLabel();
    if (renderedObject instanceof Question) {
      Question question = (Question) renderedObject;
      label.setText("<html>["
          + convertScoreToString(question.getScore()) + "] "
          + question.getTitle() +
          "</html>");
      label.setIcon(IconLoader.getIcon(QUESTION_ICON_PATH));
    } else if (renderedObject instanceof Answer) {
      Answer answer = (Answer) renderedObject;
      label.setText("<html>[" + convertScoreToString(answer.getScore()) + "] answer </html>");
      label.setIcon(IconLoader.getIcon(ANSWER_ICON_PATH));
    } else if (renderedObject instanceof Comment) {
      Comment comment = (Comment) renderedObject;
      label.setText("<html>[" + convertScoreToString(comment.getScore()) + "] comment </html>");
      label.setIcon(IconLoader.getIcon(ANSWER_ICON_PATH));
    }
    return label;
  }

  @VisibleForTesting
  static String convertScoreToString(long score) {
    NumberFormat numberFormat = new DecimalFormat("+#;-#");
    return numberFormat.format(score);
  }
}

File: src/main/java/ru/spb/se/contexthelper/ui/StackExchangeThreadsTreeCellRender.kt
package ru.spb.se.contexthelper.ui

import com.google.code.stackexchange.schema.Answer
import com.google.code.stackexchange.schema.Comment
import com.google.code.stackexchange.schema.Question
import com.intellij.openapi.util.IconLoader
import java.awt.Component
import java.text.DecimalFormat
import javax.swing.JLabel
import javax.swing.JTree
import javax.swing.tree.TreeCellRenderer

/** Renders StackExchange elements inside JTree. */
class StackExchangeThreadsTreeCellRender: TreeCellRenderer {
    override fun getTreeCellRendererComponent(
        tree: JTree?, value: Any?,
        selected: Boolean, expanded: Boolean, leaf: Boolean, row: Int, hasFocus: Boolean
    ): Component {
        val label = JLabel()
        when (value) {
            is Question -> {
                label.text = "<html>[${convertScoreToString(value.score)}] ${value.title}</html>"
                label.icon = IconLoader.getIcon(QUESTION_ICON_PATH)
            }
            is Answer -> {
                label.text = "<html>[${convertScoreToString(value.score)}] Answer </html>"
                label.icon = IconLoader.getIcon(
                    if (value.isAccepted) ACCEPTED_ANSWER_ICON_PATH else ANSWER_ICON_PATH)
            }
            is Comment -> {
                label.text = "<html>[${convertScoreToString(value.score)}] Comment </html>"
                label.icon = IconLoader.getIcon(COMMENT_ICON_PATH)
            }
        }
        return label
    }

    companion object {
        private const val QUESTION_ICON_PATH = "/icons/so-icon.png"
        private const val ACCEPTED_ANSWER_ICON_PATH = "/icons/answer-accepted.png"
        private const val ANSWER_ICON_PATH = "/icons/answer.png"
        private const val COMMENT_ICON_PATH = "/icons/comment.png"

        fun convertScoreToString(score: Long): String {
            val numberFormat = DecimalFormat("+#;-#")
            return numberFormat.format(score)
        }
    }
}